File: src/types/Game.ts
export interface Game {
  id: string;
  name: string;
  summary: GameSummary;
  cover_art: Image;
  narrative: Narrative;
  scenes: Scene[];
  characters: Character[];
  items: Item[];
  title_music: Music;
}

export interface Item {
  name: string;
  description: string;
  image: Image;
}

export interface Character {
  name: string;
  physical_description: string;
  personality: string;
  backstory: string;
  thoughts: string;
  inventory: string[];
  image: Image;
}

export interface Scene {
  name: string;
  narrative: string;
  metadata: string;
  characters: string[];
  items: string[];
  image: Image;
}

export interface Image {
  src: string;
  alt: string;
}

export interface Music {
  src: string;
  metadata: {
    index: number;
    filename: string;
    keywords: string;
    attribution: {
      title: string;
      author: string;
      from: string;
    };
  };
}

export interface GameSummary {
  name: string;
  description: string;
  artStyle: string;
  artTheme: string;
  cover_art: string;
  summary: string;
  winCondition: string;
}

export interface Narrative {
  pages: {
    narrative: string;
    image: Image;
  }[];
}

// export interface PlayerAttribute {
//   name: string;
//   description: string;
//   advancement: string;
//   significance: string;
//   startingValue: number;
// }

// export interface KeyItem {
//   name: string;
//   description: string;
//   location: string;
//   significance: string;
//   image: string;
// }

// export interface KeyArea {
//   name: string;
//   description: string;
//   significance: string;
// }

// export interface KeyCharacter {
//   name: string;
//   description: string;
//   significance: string;
//   image: string;
// }

File: src-tauri/src/game/game.rs
use serde::{Deserialize, Serialize};

use crate::file_manager::FileManager;

use super::{summary::Summary, title_music::TitleMusic, Character, Image, Item, Narrative, Scene};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Game {
    pub id: String,
    pub name: String,
    pub summary: Summary,
    pub cover_art: Image,
    pub narrative: Narrative,
    pub scenes: Vec<Scene>,
    pub characters: Vec<Character>,
    pub items: Vec<Item>,
    pub title_music: TitleMusic,
}

impl Game {
    pub fn load(
        game_id: impl Into<String>,
        file_manager: &FileManager,
    ) -> Result<Self, anyhow::Error> {
        file_manager.read_json::<Game>(format!("{}/game.json", game_id.into()))
    }
}
